### PR TITLE
Feat: add acr ee support

### LINKED ISSUE #19659: Don't support to sync image to ACR EE repositry
How can we help you?

1.  add ACR EE support

### PR DESCRIPTION
**Pull request by njucjc**
Thank you for contributing to Harbor!

# Comprehensive Summary of your change
Feat: add acr ee support

# Issue being fixed
Fixes #19659 

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [x] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

**Comment by njucjc** [comment]
@AllForNothing @chlins @MinerYang @ctrlaltdel PTAL

**Comment by wy65701436** [comment]
Thanks, @njucjc for your contribution. 

I have a couple of basic questions: 

- Is it possible to set up a staging environment of ACR EE for our validation and regression purposes? 
- Also, considering that the replication adapter ships with harbor-core, which requires CVE and bug fixes, would you be responsible for maintaining this adapter?

**Comment by njucjc** [comment]
> Thanks, @njucjc for your contribution.
> 
> I have a couple of basic questions:
> 
> * Is it possible to set up a staging environment of ACR EE for our validation and regression purposes?
> * Also, considering that the replication adapter ships with harbor-core, which requires CVE and bug fixes, would you be responsible for maintaining this adapter?

- Of course, I can provide an ACR EE instance for regression verification. But how should I provide this ACR EE instance?
- Sure, It's my honor to maintain this adapter

**Comment by njucjc** [comment]
any comments?

**Comment by njucjc** [comment]
> > Thanks, @njucjc for your contribution.
> > I have a couple of basic questions:
> > 
> > * Is it possible to set up a staging environment of ACR EE for our validation and regression purposes?
> > * Also, considering that the replication adapter ships with harbor-core, which requires CVE and bug fixes, would you be responsible for maintaining this adapter?
> 
> @wy65701436
> 
> * Of course, I can provide an ACR EE instance for regression verification. What is the best way for me to provide this ACR EE instance for you to verify？it has been verified by myself.
> * Sure, It's my honor to maintain this adapter.
> 
> By the way, I fix some `UTTEST` errors

@wy65701436 I send you an ACR EE acount for you in email !

**Comment by njucjc** [comment]
@wy65701436  maybe reopen it?

**Comment by OrlinVasilev** [comment]
just a reminder, as discussed please add test cases :)

**Comment by njucjc** [comment]
> just a reminder, as discussed please add test cases :)

I will add ut as soon as possible

File: src/pkg/reg/adapter/aliacr/adapter.go [hunk 1]
```diff
@@ -45,48 +41,80 @@ func init() {
...
 // example:
-// https://registry.%s.aliyuncs.com
 // https://cr.%s.aliyuncs.com
-// https://registry-vpc.%s.aliyuncs.com
-// https://registry-internal.%s.aliyuncs.com
-var regRegion = regexp.MustCompile(`https://(registry|cr|registry-vpc|registry-internal)\.([\w\-]+)\.aliyuncs\.com`)
+var regACRServiceURL = regexp.MustCompile(`https://cr\.([\w\-]+)\.aliyuncs\.com`)
 
-func getRegion(url string) (region string, err error) {
+func getURL(url string) (string, error) {
```
Comment (chlins) [inline src/pkg/reg/adapter/aliacr/adapter.go:47]: `getRegistryURL` may be more clearer.

Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/adapter.go]: done

**Review by chlins — APPROVED** [review]
lgtm

**Comment by njucjc** [comment]
> lgtm

@chlins  fix some ut error.

File: src/pkg/reg/adapter/aliacr/openapi_test.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	"time"
+
+	"github.com/goharbor/harbor/src/pkg/reg/model"
+)
+
+var acrAdapter *adapter
+var acreeAdapter *adapter
+
+func init() {
+	accessKey := os.Getenv("ALIYUN_ACCESS_KEY")
```
Comment (kofj) [inline src/pkg/reg/adapter/aliacr/openapi_test.go:30]: Looking at the code, we need to use the real Alibaba Cloud AK SK to run UTs. Can it be mock?

**Comment by kofj** [comment]
@njucjc I watched the demo recording of the community meeting on July 24, 2024. Can you verify that the personal version of ACR is functioning properly at the same time?

**Comment by njucjc** [comment]
> @njucjc I watched the demo recording of the community meeting on July 24, 2024. Can you verify that the personal version of ACR is functioning properly at the same time?

@kofj Yes,  the personal version of ACR can be working well at the same time.

File: src/pkg/reg/adapter/aliacr/openapi_test.go (hunk 2 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/openapi_test.go:30]: > Looking at the code, we need to use the real Alibaba Cloud AK SK to run UTs. Can it be mock?

@kofj Mock openapi seems not to have achieved the purpose of testing it. Do you have any suggestions for this part? My first thought was to provide an AK/SK to assist testing.

**Comment by chlins** [comment]
@kofj Hi, do you have any other comments for this PR?

File: src/pkg/reg/adapter/aliacr/openapi.go [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+			return nil, err
+		}
+		for _, repo := range reposResp.Repositories {
+			repos = append(repos, &repository{
+				Name:      repo.RepoName,
+				Namespace: repo.RepoNamespaceName,
+				ID:        repo.RepoId,
+			})
+		}
+		if !reposResp.ListRepositoryIsSuccess {
```
Comment (wy65701436) [inline src/pkg/reg/adapter/aliacr/openapi.go:252]: Moved the check for success right after the response.

File: src/pkg/reg/adapter/aliacr/openapi.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+	}, nil
+}
+
+// ListNamespace returns a list of namespaces
+func (acree *acreeOpenapi) ListNamespace() ([]string, error) {
+	var namespaces []string
+	var nsReq = cr_ee.CreateListNamespaceRequest()
+	nsReq.SetDomain(acree.domain)
+	nsReq.InstanceId = acree.instanceID
+	var page = 1
```
Comment (wy65701436) [inline src/pkg/reg/adapter/aliacr/openapi.go:205]: page:=1

File: src/pkg/reg/adapter/aliacr/openapi.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+// ListRepository returns a list of repositories in the specified namespace
+func (acree *acreeOpenapi) ListRepository(namespaceName string) ([]*repository, error) {
+	var repos []*repository
+	var reposReq = cr_ee.CreateListRepositoryRequest()
+	reposReq.SetDomain(acree.domain)
+	reposReq.InstanceId = acree.instanceID
+	reposReq.RepoNamespaceName = namespaceName
+	reposReq.RepoStatus = "NORMAL"
+	var page = 1
```
Comment (wy65701436) [inline src/pkg/reg/adapter/aliacr/openapi.go:238]: ditto

File: src/pkg/reg/adapter/aliacr/openapi.go [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+			})
+		}
+		if !reposResp.ListRepositoryIsSuccess {
+			return nil, fmt.Errorf("failed to list repo: %s", reposResp.GetHttpContentString())
+		}
+		total, err := strconv.Atoi(reposResp.TotalCount)
+		if err != nil {
+			return nil, err
+		}
+		if total-(reposResp.PageNo*reposResp.PageSize) <= 0 {
```
Comment (wy65701436) [inline src/pkg/reg/adapter/aliacr/openapi.go:259]:
```
if total <= page*reposResp.PageSize {
     break
}
```

File: src/pkg/reg/adapter/aliacr/openapi.go [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+	return &acrOpenapi{
+		client: client,
+		domain: fmt.Sprintf(endpointTpl, regionID),
+	}, nil
+}
+
+// ListNamespace returns a list of namespaces
+func (acr *acrOpenapi) ListNamespace() ([]string, error) {
+	var namespaces []string
+	var nsReq = cr.CreateGetNamespaceListRequest()
```
Comment (wy65701436) [inline src/pkg/reg/adapter/aliacr/openapi.go:74]: Please revise them a little bit to make it more readable.

```
nsReq := cr.CreateGetNamespaceListRequest()
nsReq.SetDomain(acr.domain)

nsResp, err := acr.client.GetNamespaceList(nsReq)
if err != nil {
	return nil, err
}
```

File: src/pkg/reg/adapter/aliacr/adapter.go [hunk 8]
```diff
@@ -45,48 +41,80 @@ func init() {
...
+	return fmt.Sprintf(registryEndpointTpl, rs[1]), nil
+}
+
+// example:
+// registry.aliyuncs.com:cn-hangzhou:china:cri-xxxxxxxxx
+// registry.aliyuncs.com:cn-hangzhou:26842
+func parseRegistryService(service string) (*registryServiceInfo, error) {
+	parts := strings.Split(service, ":")
+	length := len(parts)
+	if length <= 0 {
```
Comment (wy65701436) [inline src/pkg/reg/adapter/aliacr/adapter.go:64]: if length < 2 {
	return nil, errors.New("invalid service format: expected 'registry.aliyuncs.com:region:xxxxx'")
}

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/pkg/reg/adapter/aliacr/openapi.go (hunk 3 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/openapi.go]: done

File: src/pkg/reg/adapter/aliacr/openapi.go (hunk 4 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/openapi.go]: done

File: src/pkg/reg/adapter/aliacr/openapi.go (hunk 5 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/openapi.go]: done

File: src/pkg/reg/adapter/aliacr/openapi.go (hunk 6 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/openapi.go]: done

File: src/pkg/reg/adapter/aliacr/openapi.go (hunk 7 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/openapi.go]: done

File: src/pkg/reg/adapter/aliacr/adapter.go (hunk 8 above)
Comment (njucjc) [reply src/pkg/reg/adapter/aliacr/adapter.go]: done

**Pull request merged**